lottie-compose/src/main/java/com/airbnb/lottie/compose/animateLottieCompositionSuspend.kt

```diff
@@ -0,0 +1,83 @@
+package com.airbnb.lottie.compose
+
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.withFrameNanos
+import com.airbnb.lottie.LottieComposition
+import kotlinx.coroutines.NonCancellable
+import kotlinx.coroutines.withContext
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.EmptyCoroutineContext
+
+/**
+ * Determines how the animation should behave if the calling CoroutineScope is cancelled.
+ */
+enum class LottieCancellationBehavior {
+    /**
+     * Stop animation immediately and return early.
+     *
+     * @see lottieTransition
+     */
+    Immediate,
+
+    /**
+     * Continue suspending until the animation completes, effectively ignoring the cancellation request.
+     * This may be useful if you have a segmented animation with multiple states and you want them to
+     * animate smoothly from one to the next when the state changes.
+     *
+     * @see lottieTransition
+     */
+    AtEnd,
+}
+
+/**
+ * This is similar to [animateLottieComposition] which returns progress as a MutableState<Float>.
+ * However, instead of returning state, this takes [progress] as a parameter which it updates
+ * internally and suspends until the animation completes.
+ *
+ * To repeat an animation multiple times, just wrap this invocation in a while loop.
+ *
+ * If the calling CoroutineScope is cancelled, [cancellationBehavior] determines how cancellation
+ * should happen. [LottieCancellationBehavior.AtEnd] will ignore the cancellation and continue suspending
+ * until the animation completes. [LottieCancellationBehavior.Immediate] will immediately cancel the
+ * animation and return early.
+ *
+ * @see lottieTransition
+ */
+suspend fun animateLottieComposition(
+    composition: LottieComposition?,
+    progress: MutableState<Float>,
```

Comment:
@lelandrichardson is it okay for this to take `MutableState<Float>` or should it be a `MutableStateFlow`? The semantics of this API works nicely with `MutableState` but it is rare to see a third party library use `MutableState` in a public API.

Reply:
This `MutableState` seems to have two sources of truth. I'm guessing the intention here is to support seeking while playing. Since when `progress` is mutated outside of this suspend fun, it effectively resets the initialProgress of the animation, I wonder if it would more clear to have separate concepts for initialProgress vs. animatedProgress. 

BTW, what's the recommended way seek with gesture?

Reply:
@doris4lt you are correct. It does, in fact, have 2 sources of truth so you can adjust the progress while it animates. When you do that, it snaps to that progress within the current repeatCount. This is how we achieve playing + dragging the seek bar in the sample app. Does this seem like a reasonable approach to do that?

Reply:
The multiple sources of truth may be error-prone, as it's hard to track down a bug when something goes wrong. At the same time, it's limiting - using `MutableState` here only allows devs to change progress. If they were to imperatively control other behaviors of the animation, it needs to be done at an entirely separate place. I can imagine scenarios where when changing the progress with a gesture, devs may prefer the animation to not auto-play until finger release.

I would create a set of imperative `suspend fun` APIs for various controls that would affect animation's lifecycle, such as changing the starting progress of a run, stopping the animation, snapping to a progress without running (i.e. seek only), etc.  That way devs could control the animation in one LaunchedEffect. 

For the multiple sources of truth, I would recommend writing the current progress of the animation to a different animation state object, along with other animation related states (e.g. whether it's playing, iteration #, etc). They could read only.  I assume the reason for writing the animation progress back to the MutableState is so that devs could inspect the current progress of the animation and make decisions accordingly. Though the progress alone may not be sufficient for such a decision. Hence the recommendation for a separate state object altogether. :-)

Reply:
@doris4lt I think you raise a lot of valid points. However, I can't seem to get an imperative corutines API to come together in a way I'm happy with. Here are 3 branches in which I tried some different approaches:
https://github.com/airbnb/lottie-android/commits/gpeal/compose-min-max-frame-imperative-api
https://github.com/airbnb/lottie-android/commits/gpeal/compose-min-max-frame-imperative-suspend-api-2
https://github.com/airbnb/lottie-android/commits/gpeal/compose-min-max-frame-imperative-api-3

I'd love another set of eyes or ideas for an API that works. In general, the dual-ownership aspect of these animations made most of these APIs really tricky/confusing.

Do you think there are any major issues with these APIs as they are? If not, I'd be open to adding additional animation APIs in the future now that the base LottieAnimation composable just takes a raw progress float.